File: tests/color.test.js
import { test } from 'node:test';
import assert from 'node:assert';
import { hexToRgba, parseHex } from '../js/utils.js';

test('#FF0000 with alpha 0.5 → rgba(255,0,0,0.5)', () => {
    assert.strictEqual(hexToRgba('#FF0000', 0.5), 'rgba(255,0,0,0.5)');
});

test('lowercase #00ff00 with alpha 1 → rgba(0,255,0,1)', () => {
    assert.strictEqual(hexToRgba('#00ff00', 1), 'rgba(0,255,0,1)');
});

test('pass-through for named color "transparent"', () => {
    assert.strictEqual(hexToRgba('transparent', 0.5), 'transparent');
});

test('pass-through for existing rgba string', () => {
    assert.strictEqual(hexToRgba('rgba(1,2,3,0.4)', 0.5), 'rgba(1,2,3,0.4)');
});

test('pass-through for short hex', () => {
    assert.strictEqual(hexToRgba('#fff', 0.5), '#fff');
});

test('pass-through for non-string', () => {
    assert.strictEqual(hexToRgba(123, 0.5), 123);
});

test('parseHex(#3399ff) → {r:51,g:153,b:255}', () => {
    assert.deepStrictEqual(parseHex('#3399ff'), { r: 51, g: 153, b: 255 });
});

File: core/tests/color.test.cpp
#include "doctest.h"
#include "color.hpp"

using namespace stencil::core;

// Mirrors browser/tests/color.test.js.

TEST_CASE("#FF0000 with alpha 0.5 -> rgba(255,0,0,0.5)") {
  CHECK(hexToRgba("#FF0000", 0.5) == "rgba(255,0,0,0.5)");
}

TEST_CASE("lowercase #00ff00 with alpha 1 -> rgba(0,255,0,1)") {
  CHECK(hexToRgba("#00ff00", 1) == "rgba(0,255,0,1)");
}

TEST_CASE("pass-through for named color \"transparent\"") {
  CHECK(hexToRgba("transparent", 0.5) == "transparent");
}

TEST_CASE("pass-through for existing rgba string") {
  CHECK(hexToRgba("rgba(1,2,3,0.4)", 0.5) == "rgba(1,2,3,0.4)");
}

TEST_CASE("pass-through for short hex") {
  CHECK(hexToRgba("#fff", 0.5) == "#fff");
}

TEST_CASE("parseHex(#3399ff) -> {51,153,255}") {
  const auto rgb = parseHex("#3399ff");
  REQUIRE(rgb.has_value());
  CHECK(rgb->r == 51);
  CHECK(rgb->g == 153);
  CHECK(rgb->b == 255);
}

TEST_CASE("parseHex rejects non-hex / short input") {
  CHECK_FALSE(parseHex("#fff").has_value());
  CHECK_FALSE(parseHex("transparent").has_value());
  CHECK_FALSE(parseHex("#gggggg").has_value());
}
